src/main/java/hudson/remoting/PingThread.java

```diff
@@ -120,9 +120,10 @@ private void ping() throws IOException, InterruptedException {
                 // get method waits "at most the amount specified in the timeout",
                 // so let's make sure that it really waited enough
             }
+            remaining = end - System.nanoTime();
         } while(remaining>0);
 
-        onDead(new TimeoutException("Ping started on "+start+" hasn't completed at "+System.currentTimeMillis()));//.initCause(e)
+        onDead(new TimeoutException("Ping started at "+start+" hasn't completed by "+System.currentTimeMillis()));//.initCause(e)
```

Comment:
Not related to this PR, but probably this should use time formats, not raw `long`s. (Most easily: `new Date(…)`)